File: 112/path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        if root is None:
            return False

        if root.left is None and root.right is None:
            if root.val == sum:
                return True
            return False

        has_left_path, has_right_path = False, False
        if root.left is not None:
            has_left_path = self.hasPathSum(root.left, sum - root.val)

        if root.right is not None:
            has_right_path = self.hasPathSum(root.right, sum - root.val)

        return has_left_path or has_right_path

File: 112/PathSum.scala
object PathSum {
//  def hasPathSum(root: TreeNode, sum: Int): Boolean = {
//    if (null == root) return false
//    if (null == root.left && null == root.right) {
//      if (root.value == sum) return true
//      else return false
//    }
//
//    var hasLeftPath: Boolean = false
//    var hasRightPath: Boolean = false
//    if (root.left != null) hasLeftPath = hasPathSum(root.left, sum - root.value)
//    if (root.right != null) hasRightPath = hasPathSum(root.right, sum - root.value)
//    return hasLeftPath || hasRightPath
//  }

  def hasPathSum(root: TreeNode, sum: Int): Boolean = {
    def loop(root: TreeNode, target: Int): Boolean =
      if (root == null) false
      else if (root.left == null && root.right == null) target + root.value == sum
      else loop(root.left, target + root.value) || loop(root.right, target + root.value)

    loop(root, 0)
  }

  class TreeNode(var _value: Int = 0) {
    var value: Int = _value
    var left: TreeNode = null
    var right: TreeNode = null
  }
}
